README
Byterun
-------

This is a pure-Python implementation of a Python bytecode execution virtual
machine.  I started it to get a better understanding of bytecodes so I could
fix branch coverage bugs in coverage.py.



NEWS.md
1.5.0 2021-11-25
================

* Better PyPy 3.6-3.8 opcode coverages,
* Better 3.9 opcode coverage
* Start handling Python 3.10
* Add bytecode files and testing for the above.

1.4.0 2021-11-07
================

* Use newer xdis which supports 3.10. We don't have 3.10 opcode interpretation yet though.
  This was a lot of upheaval since we can no longer use floats to represent python versions:
  3.1 == 3.10 in floating point. As xdis uses, we use tuples now.
* Improve fake `version_info` for PyPy.
* Start pypy37 and pypy38 bytecodes, although this is not finished yet. You interpret other bytecode from PyPy though.
* Tweak `FORMAT_VALUE` and improve assembly display

1.3.6 2021-09-04
================

* Handle Python 3.7 - 3.9 better
* fake sys.version, sys.version_info, and sys.hexversion in cross interpreting
* Handle PyPy 3.5 and other PyPy's better
* Fix bug in reporting line number
* Fix bug in Python 2.{4,5} MAKE_CLOSURE
* Improve ofrmating on IMPORT_NAME on 2.6 and later
* Type check VM arguments more

1.3.5 2020-07-24
================

Note: a lot of stdlib tests have become broken. This is probably ood because it reflects that we are honestly interpreting more and
not running interpreter code as builtin Python functions (e.g. eval and exec used to work this way.)

* Start interpreter assisted breakpoints.
* Fill out interpreter inspect more
* exec <bytes> is the same as exec <str>
* make version pydoc frendly
* propagate interpreter-caught exceptions more properly

1.3.4 2020-06-28
================

Small changes.

* `run_eval()` renamed to `eval_frame()` to better match the way CPython does it
* `exec()`, and `eval()` builtins go through `compile` and `eval_frame()` so they get interpreted
* It is okay when `exec()` and `eval()` which call `eval_frame()` leave data on the stack (?)
* `__module__` property set properly on `eval_frame()`
* `pretty_event_flags()` will show VM tracing event flags nicley. This helps [trepan-xpy](https://pypi.org/project/trepanxpy/)
* `__qualname__` and `__name__` are set differently and properly in `MAKE_FUNCTION` in 3.4+
* More tests and expanded coverage
* Work progress in getting 3.6+ `GET_AWAITABLE` done, but more work is needed.

1.3.3 2020-06-13 Fleeetwood66+
==============================

Overall, improved cross-version bytecode interpreting and some bug fixes.

* Correct `BUILD_MAP_UNPACK` opcode handling.
* Set namespace locals and globals to the proper namespace in builtin `eval()` like we do for `locals()` and `globals()` built-in functions
* Better built-in test for `eval()`. Note Tk has its own built-in `eval()`
* Handle `__idiv__` in overwritten insdie `//=`
* Improve argument format for `CALL_FUNCTION_KW` for 3.6+
* Initialize frame cells from closure, not `f_back`  See https://github.com/nedbat/byterun/issues/17

1.3.2 2020-06-03 Marilyn + 1
============================

Overall, some bugs were fixed especially on the 3.6+ side.

There is better cross version interpreting.

Release before we do a major overhaul on the closure or cell "freeops", e.g. `LOAD_DEREF` and `STORE_DEREF`.

* Fix bug in detecting 3.6+ keyword-only signatures
* Correct bugs in 3.6+ `MAKE_FUNCTION`
* Improve our build_class(); don't try to convert built-in `__entry__`, `__exit__` routines
* Add Python 2.4 bytecode interpreting
* Improve logging operator formatting


1.3.1 2020-05-30 Lady Elaine
============================

* Logging often shows the operand that an opcode uses
* Cross-version bytecode interpreting has been improved. Adding inter-operable `inspect` modules helps here.
* Support for debugger-like "step over" and "finish" trace commands was added.

1.3.0 was a botched release. xpython.stdlib wasn't included

1.2.1 2020-04-20 trepan-xpy
===========================

The main thrust of this release was to make it suitable to be used from a debugger.

* There is better (but not exact) conformance of Python's `sys.settrace()` API:
   - linestarts is a frame-oriented thing, not a VM thing.
   - Store callback in `f_trace`.
* Allow supplying a callback print function, so that the client can colorize the parts of the instruction
* Add screencasts


1.2.0 2020-04-19 Primidi 1st Prairial - Alfalfa - HF
====================================================

This version adds callback hooks for tracing or debugger that will be released soon.

There is more complete (but not fully complete) Version 3.5-3.7 bytecode interpretation.
(`BUILD_TUPLE_UNPACK_WITH_CALL` implemented 3.6 `MAKE_FUNCTION` corrected.)

`build_class()` has been fixed so that it picks up class variables. This means we do better at as a cross version interpreter - that's where `build_class()` is currently used.

Via `build_class()` we can track into `__entry__()` and `__exit__()` functions of a context manager.

`frame.f_lasti` had been pointing to the instruction that might be run next rather than the one that is about to be run. While this simplified interpreter implementation, it didn't follow CPython's meaning. And when this is used from a debugger this becomes unusable.

Disassembly output has been fixed up courtesy of argument information from `xdis`. `xdis` imports have been simplified in version 4.6.0, so we require that.

Additional PyPy bytecode ops (`LOOKUP_METHOD`, `CALL_METHOD`) are supported; so you can run this from PyPy now or do cross-interpretation from the corresponding CPython bytecode.

Better conformance of Python's Frame type. Added: `__qualname__` and set `__annotation__` attributes.



1.1.0 2020-05-08 Lewis-10
=========================

Here we support up to 3.6 and 3.7 now -- we're the first on the block to support Python 3.7! PyPy versions 2.7, 3.2, 3.5, and 3.6 are also supported.

A number of bug have been squashed but others still run rampant. Some the lesser-used 3.6-3.7 and PyPy opcodes haven't been implemented. However about 300 of the tests that Python's uses to test itself pass; the same is true for other versions.


__pkginfo__.py
import sys

author = "Rocky Bernstein, Ned Batchelder, Paul Swartz, Allison Kaptur and others"
author_email = "rb@dustyfeet.com"
entry_points = {"console_scripts": ["xpython=xpython.__main__:main"]}

# Python-version | package | last-version |
# -----------------------------------------
# 3.2            | click   | 4.0          |
# 3.2            | pip     | 8.1.2        |
# 3.3            | pip     | 10.0.1       |
# 3.4            | pip     | 19.1.1       |

import os.path as osp


def get_srcdir():
    filename = osp.normcase(osp.dirname(osp.abspath(__file__)))
    return osp.realpath(filename)


def read(*rnames):
    return open(osp.join(get_srcdir(), *rnames)).read()


exec(read("xpython/version.py"))
exec(read("xpython/version_info.py"))

# Don't assume we have xdis installed.
PYTHON_VERSION_TRIPLE = tuple(sys.version_info[:3])

IS_PYPY = "__pypy__" in sys.builtin_module_names

supported_versions = SUPPORTED_PYPY if IS_PYPY else SUPPORTED_PYTHON  # noqa
mess = "PYPY 2.7, 3.2, 3.5-3.7" if IS_PYPY else "CPython 2.7, 3.2 .. 3.10"

if PYTHON_VERSION_TRIPLE[:2] not in supported_versions:
    python = "PyPy " if IS_PYPY else "C"
    print("We only support %s; you have %sPython %s" % (mess, python, sys.version_info))
    raise Exception(mess)

if (3, 0) <= PYTHON_VERSION_TRIPLE < (3, 3):
    click_version = "<= 4.0"
else:
    click_version = ""

install_requires = (["six", "xdis >= 6.0.3, < 6.1.0", "click%s" % click_version],)

py_modules = None
short_desc = "Python cross-version byte-code interpeter"
url = "http://github.com/rocky/xpython"

classifiers = [
    "Operating System :: OS Independent",
    "Programming Language :: Python :: 2.7",
    "Programming Language :: Python :: 3.2",
    "Programming Language :: Python :: 3.3",
    "Programming Language :: Python :: 3.4",
    "Programming Language :: Python :: 3.5",
    "Programming Language :: Python :: 3.6",
    "Programming Language :: Python :: 3.7",
    "Programming Language :: Python :: 3.8",
    "Programming Language :: Python :: 3.9",
    "Programming Language :: Python :: 3.10",
    "Programming Language :: Python :: Implementation :: PyPy",
]

# Get/set VERSION and long_description from files
long_description = read("README.rst") + "\n"
exec(read("xpython/version.py"))
